# 10장 심화문제
- 윤주향
- 201841089

# 1.중요사항 정리   
- 강의에서 가장 중요하고 생각하는 용어, 개념 등을 세 개 이상 써주세요.   
   1) 넘파이(Numpy) : 대용량의 배열과 행렬연산을 빠르게 수행하고, 고차원적인 수학 연산자와 함수를 포함하고 있는 파이썬 라이브러리이다.     
   2) 넘파이의 다차원배열(ndarray)는 프로그램의 오류를 찾거나 배열의 상세한 정보를 손쉽게 조회할 수 잇다.     
   3) 넘파이 배열에는 +연산자나 * 연산자와 같은 수학적인 연산자를 얼마든지 적용할 수 있다.     
      


# 2. 자기 점검

   - 강의 중 어려웠던 점이나, 잘 모르겠는 것들을 써주세요.  
     
     문제 10.3을 푸는데 문제에서 랜덤시드가 주어지지 않아 랜덤시드를 100으로 주고 풀었는데, 이렇게 풀어도 되나요??    
   

# 3. 심화문제 풀기

   - 10단원의 심화 문제를 풀어서 제출하세요.

# 10.1

In [2]:
import numpy as np
num_arr = np.array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20])
print(num_arr)
print(num_arr[::-1])
print("num_arr 내의 모든 원소의 합 : ", sum(num_arr))
print(num_arr.reshape(5, 4))

[ 1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20]
[20 19 18 17 16 15 14 13 12 11 10  9  8  7  6  5  4  3  2  1]
num_arr 내의 모든 원소의 합 :  210
[[ 1  2  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]
 [13 14 15 16]
 [17 18 19 20]]


# 10.2

In [36]:
#1)
n = np.arange(25)
n_arr = n.reshape(5,5)
print(n_arr)

[[ 0  1  2  3  4]
 [ 5  6  7  8  9]
 [10 11 12 13 14]
 [15 16 17 18 19]
 [20 21 22 23 24]]


In [13]:
#2)
print("첫 원소 : {0} \n마지막 원소 : {1}".format(n_arr[0,0],n_arr[4,-1]))

첫 원소 : 0 
마지막 원소 : 24


In [37]:
#3)
print(n_arr[0:2,::1])

[[0 1 2 3 4]
 [5 6 7 8 9]]


In [38]:
#4)
print(n_arr[2::, ::1])

[[10 11 12 13 14]
 [15 16 17 18 19]
 [20 21 22 23 24]]


In [39]:
#5)
print(n_arr[::1, ::2])

[[ 0  2  4]
 [ 5  7  9]
 [10 12 14]
 [15 17 19]
 [20 22 24]]


In [40]:
#6)
print(n_arr[::2, ::2])

[[ 0  2  4]
 [10 12 14]
 [20 22 24]]


In [41]:
#7)
s = n_arr[0:2,::1]
print(s.reshape(5,2))

[[0 1]
 [2 3]
 [4 5]
 [6 7]
 [8 9]]


# 10.3

In [47]:
#1)
np.random.seed(100) #랜덤시드를 임의의 수인 100으로 지정했다.
a = np.random.rand(3,3,3)
print(a)

[[[0.54340494 0.27836939 0.42451759]
  [0.84477613 0.00471886 0.12156912]
  [0.67074908 0.82585276 0.13670659]]

 [[0.57509333 0.89132195 0.20920212]
  [0.18532822 0.10837689 0.21969749]
  [0.97862378 0.81168315 0.17194101]]

 [[0.81622475 0.27407375 0.43170418]
  [0.94002982 0.81764938 0.33611195]
  [0.17541045 0.37283205 0.00568851]]]


In [48]:
#2)
print("a의 원소들 중 최댓값 : ",np.max(a))

a의 원소들 중 최댓값 :  0.9786237847073697


In [49]:
#3)
print("a의 원소들 중 최댓값의 위치 : ",np.argmax(a))

a의 원소들 중 최댓값의 위치 :  15


# 10.4

In [104]:
#1)
x = np.zeros((5,5), dtype = int) #모두 0인 5*5배열
x[::2, ::2] = 1 
x[1::2,1::2] = 1 #해당위치에 직접 1을 넣어서 생성한다.
print(x)

[[1 0 1 0 1]
 [0 1 0 1 0]
 [1 0 1 0 1]
 [0 1 0 1 0]
 [1 0 1 0 1]]


In [107]:
#2)
print("행렬의 행 방향 성분의 합 :")
arr = [sum(x[i]) for i in range(0,5)]
print(arr)

행렬의 행 방향 성분의 합 :
[3, 2, 3, 2, 3]


# 10.5

In [65]:
a = np.arange(0,32).reshape(4,4,2)

In [64]:
b = a.flatten() #flatten()함수는 여러 행렬의 함수를 하나의 행 또는 열로 변환해준다. 
print("10번째 원소 : {}\n20번째 원소 : {}".format(b[9],b[19]))

10번째 원소 : 9
20번째 원소 : 19


# 10.6

In [72]:
x1 = [i for i in range(100)]
x2 = [i + np.random.randint(1,10) for i in range(100)]
x3 = [i + np.random.randint(1,50) for i in range(100)]
x4 = [np.random.randint(1,100) for i in range(100)]

In [76]:
result = np.corrcoef( [x1, x2, x3, x4] )
print(result)

[[ 1.          0.99605685  0.87687476  0.01575329]
 [ 0.99605685  1.          0.8678746  -0.00244552]
 [ 0.87687476  0.8678746   1.          0.08084923]
 [ 0.01575329 -0.00244552  0.08084923  1.        ]]


상관관계   
x1과 x2 : 0.9961   
x1과 x3 : 0.8769   
x1과 x4 ; 0.0158   

### 1) x1과 가장 상관관계가 높은 리스트는 무엇인가? x2   
### 2) x1과 가장 상관관계가 낮은 리스트는 무엇인가? x4   